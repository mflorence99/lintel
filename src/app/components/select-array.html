<ng-container
  [formGroup]="selectArrayForm">

  <div 
    [ngStyle]="{'grid-template-columns': 'repeat(auto-fit, ' + columnWidth + ')'}"
    class="selects">

    <lintel-singleselector
      [formControl]="select"
      [options]="options"
      *ngFor="let select of selectArrayForm.controls.selects['controls']; let ix = index">

      <fa-icon 
        (click)="removeSelector(ix)"
        [icon]="['fas', 'times']"
        *ngIf="selectArrayForm.controls.selects['controls'].length > 1"
        suffix
        title="Remove this item">
      </fa-icon>

      <a
        (click)="addSelector()"
        *ngIf="ix === (selectArrayForm.controls.selects['controls'].length - 1) && ix < (maxItems - 1)"
        suffix>
        More
      </a>

    </lintel-singleselector>

  </div>

</ng-container>
