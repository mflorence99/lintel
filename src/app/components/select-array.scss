:host {
  align-items: center;
  display: grid;
}

.selects {
  display: grid;
  grid-gap: 0.5rem;
  // NOTE: actual with set by style from control input
  // grid-template-columns: repeat(auto-fit, ...);

  .disabled {
    cursor: not-allowed;
    opacity: var(--lintel-opacity-disabled);
  }
  
}
