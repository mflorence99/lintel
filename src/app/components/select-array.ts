import { DestroyService } from '../services/destroy';
import { SingleselectorOptions } from './singleselector';

import { ChangeDetectionStrategy } from '@angular/core';
import { ChangeDetectorRef } from '@angular/core';
import { Component } from '@angular/core';
import { ControlValueAccessor } from '@angular/forms';
import { FormArray } from '@angular/forms';
import { FormBuilder } from '@angular/forms';
import { FormControl } from '@angular/forms';
import { FormGroup } from '@angular/forms';
import { Input } from '@angular/core';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { OnInit } from '@angular/core';

import { filter } from 'rxjs/operators';
import { forwardRef } from '@angular/core';
import { takeUntil } from 'rxjs/operators';

export type SelectArrayType = (number | string)[];

/**
 * Select array form control
 *
 * NOTE: just enough to be able to match VSCode as well as possible
 *
 * @see https://blog.thoughtram.io/angular/2016/07/27/custom-form-controls-in-angular-2.html
 * 
 * Reference is old, but still helpful
 */

@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  providers: [
    {
      provide: NG_VALUE_ACCESSOR,
      useExisting: forwardRef(() => SelectArrayComponent),
      multi: true
    },
    DestroyService
  ],
  selector: 'lintel-select-array',
  templateUrl: 'select-array.html',
  styleUrls: ['select-array.scss']
})

export class SelectArrayComponent implements ControlValueAccessor, OnInit {

  @Input() columnWidth = '10rem';
  @Input() enabled = true;
  @Input() maxItems = Infinity;
  @Input() options: SingleselectorOptions = [];

  selectArrayForm: FormGroup;

  @Input() type: 'number' | 'text' = 'text';
  @Input() uniqueItems: boolean;

  values: SelectArrayType = [];

  @Input()
  get value(): SelectArrayType {
    return this.values;
  }
  set value(value: SelectArrayType) {
    this.values = value ?? [];
    this.underConstruction = true;
    const selects = this.selectArrayForm.controls.selects as FormArray;
    // NOTE: there's always one more input for a new value
    // trim off excess selects
    while (selects.length > (this.values.length + 1))
      selects.removeAt(selects.length - 1);
    while (selects.length < (this.values.length + 1))
      selects.push(new FormControl(null));
    // patch in the new values
    selects.patchValue([...this.values, null], { emitEvent: false });
    this.underConstruction = false;
    this.onChange?.(this.values);
    // TODO: Angular can be so weird!
    this.cdf.detectChanges();
  }

  private onChange: Function;
  private underConstruction: boolean;

  /** ctor */
  constructor(private cdf: ChangeDetectorRef,
              private destroy$: DestroyService,
              private formBuilder: FormBuilder) {
    // initialize the form
    this.selectArrayForm = this.formBuilder.group({
      selects: new FormArray([])
    });
  }

  /** Add another selector */
  addSelector(): void {
    const selects = this.selectArrayForm.controls.selects as FormArray;
    selects.push(new FormControl(null));
  }

  /** When we're ready */
  ngOnInit(): void {
    this.selectArrayForm.valueChanges
      .pipe(
        filter(_ => !this.underConstruction),
        takeUntil(this.destroy$)
      )
      .subscribe(value => {
        this.values = value.selects
          .filter(val => !!val);
        if (this.uniqueItems)
          this.values = Array.from(new Set(this.values));
        this.onChange?.(this.value);
      });
  }

  /** @see ControlValueAccessor */
  registerOnChange(fn): void {
    this.onChange = fn;
  }

  /** @see ControlValueAccessor */
  registerOnTouched(_): void { }

  /** Remove specified selector */
  removeSelector(ix: number): void {
    const selects = this.selectArrayForm.controls.selects as FormArray;
    selects.removeAt(ix);
  }

  /** @see ControlValueAccessor */
  writeValue(value): void {
    this.value = value;
  }

}
