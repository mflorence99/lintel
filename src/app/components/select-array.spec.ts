import { SelectArrayComponent } from './select-array';

import { prepare } from './component.spec';

import 'jest-extended';

import { ComponentFixture } from '@angular/core/testing';
import { FormArray } from '@angular/forms';
import { TestBed } from '@angular/core/testing';

describe('SelectArrayComponent', () => {
  let component: SelectArrayComponent;
  let fixture: ComponentFixture<SelectArrayComponent>;

  beforeEach(() => {
    prepare();
    fixture = TestBed.createComponent(SelectArrayComponent);
    component = fixture.componentInstance;
    component.ngOnInit();
  });

  test('registerOnChange', () => {
    component.registerOnChange(jest.fn());
    expect(component['onChange']).toBeTruthy();
  });

  test('registerOnTouched', () => {
    expect(component.registerOnTouched()).toBeFalsy();
  });

  test('writeValue of number', () => {
    component.type = 'number';
    expect(component.value).toEqual([]);
    component.writeValue([1, 2, 3]);
    expect(component.value).toEqual([1, 2, 3]);
  });

  test('writeValue of string', () => {
    component.type = 'text';
    expect(component.value).toEqual([]);
    component.writeValue(['London', 'Paris', 'Rome']);
    expect(component.value).toEqual(['London', 'Paris', 'Rome']);
  });

  test('addSelector', () => {
    component.type = 'text';
    component.value = ['London', 'Paris', 'Rome'];
    expect(
      (component.selectArrayForm.controls.selects as FormArray).length
    ).toBe(4);
    component.addSelector();
    expect(
      (component.selectArrayForm.controls.selects as FormArray).length
    ).toBe(5);
  });

  test('removeSelector', () => {
    component.type = 'number';
    component.value = [1, 2, 3];
    expect(
      (component.selectArrayForm.controls.selects as FormArray).length
    ).toBe(4);
    component.removeSelector(1);
    expect(
      (component.selectArrayForm.controls.selects as FormArray).length
    ).toBe(3);
  });

  test('ngOnInit', (done) => {
    component.type = 'text';
    component.uniqueItems = true;
    component.value = ['London', 'Paris', 'Paris', 'Rome'];
    component.registerOnChange((value) => {
      expect(value).toEqual(['London', 'Paris', 'Rome']);
      done();
    });
    // NOTE: trips valueChanges
    component.addSelector();
  });

  test('snapshot', () => {
    component.type = 'text';
    component.uniqueItems = true;
    component.value = ['London', 'Paris', 'Paris', 'Rome'];
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
