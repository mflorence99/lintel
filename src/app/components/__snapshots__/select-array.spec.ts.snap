// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectArrayComponent snapshot 1`] = `
<lintel-select-array
  __ngContext__={[Function LRootView]}
  cdf={[Function ViewRef]}
  columnWidth={[Function String]}
  destroy$={[Function DestroyService]}
  enabled={[Function Boolean]}
  formBuilder={[Function FormBuilder]}
  maxItems={[Function Number]}
  options={[Function Array]}
  selectArrayForm={[Function FormGroup]}
  type={[Function String]}
  underConstruction="false"
  uniqueItems={[Function Boolean]}
  values={[Function Array]}
>
  <p
    class="selects"
    style="grid-template-columns: repeat(auto-fit, 10rem);"
  >
    <lintel-singleselector
      class="ng-untouched ng-pristine ng-valid"
      placeholder=" "
    >
      <select
        autocomplete="off"
        tabindex="0"
      >
        <option
          value="null"
        >
           
        </option>
        
        
      </select>
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
    </lintel-singleselector>
    <lintel-singleselector
      class="ng-untouched ng-pristine ng-valid"
      placeholder=" "
    >
      <select
        autocomplete="off"
        tabindex="0"
      >
        <option
          value="null"
        >
           
        </option>
        
        
      </select>
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
    </lintel-singleselector>
    <lintel-singleselector
      class="ng-untouched ng-pristine ng-valid"
      placeholder=" "
    >
      <select
        autocomplete="off"
        tabindex="0"
      >
        <option
          value="null"
        >
           
        </option>
        
        
      </select>
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
    </lintel-singleselector>
    <lintel-singleselector
      class="ng-untouched ng-pristine ng-valid"
      placeholder=" "
    >
      <select
        autocomplete="off"
        tabindex="0"
      >
        <option
          value="null"
        >
           
        </option>
        
        
      </select>
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      
    </lintel-singleselector>
    <lintel-singleselector
      class="ng-untouched ng-pristine ng-valid"
      placeholder=" "
    >
      <select
        autocomplete="off"
        tabindex="0"
      >
        <option
          value="null"
        >
           
        </option>
        
        
      </select>
      <fa-icon
        class="ng-fa-icon"
        suffix=""
        title="Remove this item"
      >
        <svg
          aria-labelledby="svg-inline--fa-title-xxx"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <title
            id="svg-inline--fa-title-xxx"
          >
            Remove this item
          </title>
          <path
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </fa-icon>
      
      <a
        suffix=""
      >
         More 
      </a>
      
    </lintel-singleselector>
    
  </p>
</lintel-select-array>
`;
